lib/mfg.c

```diff
@@ -300,6 +313,31 @@ int switchtec_mailbox_to_file(struct switchtec_dev *dev, int fd)
 	return 0;
 }
 
+static int convert_spi_clk_rate(float clk_float, int hi_rate)
+{
+	int i;
+	float *p;
+
+	if (hi_rate)
+		p = spi_clk_hi_rate_float;
+	else
+		p = spi_clk_rate_float;
+
+	if (clk_float >= p[0])
+		return 0;
+	else if (clk_float <= p[9])
+		return 9;
+
+	for (i = 0; i < 9; i++) {
+		if (clk_float >= (p[i] + p[i+1])/2)
+			return i;
+		else if (clk_float >= p[i+1])
+			return i+1;
```

Comment:
Why not use a string for the clock rate if we will have to lose accuracy when using float? A string is also good for printing.

Reply:
I don't know why you are worried about accuracy. If you want more accuracy, use a double; but you really don't need it.  A string might be good for printing, but it's not good for comparing or anything else.

Reply:
I'm thinking we should error out if the user input clock rate doesn't hit any one in the array, instead of picking a close one. When user want a particular rate but input an unintended wrong value, we should report the error rather than take it. Even using float, we can compare it with each one in the array within the accuracy provided by float, rather than choose a close one. I believe user would want to have errors reported when the input is not a valid value, since this is an OTP thing. So why not use more accurate type like string or enum? 

It's not hard to compare the input string with an array of valid values, just like what we're doing here for the float. If string is not ideal, how about using enum and a helper in the library to map it to a string for printing?

Reply:
Generally the user doesn't really care about the exact rate. They only really need to specify a maximum.  In which case we should probably ensure that the we don't set it higher than the user requests.

Reply:
That might be true for some other setting operation. However, I still believe we should set the exact rate user requested for this one time programming operation, as any mistake would cause unrecoverable damage to the chip (and the system). User should be responsible for the OTP setting values, and we should execute exactly what user asks us to do.

Reply:
It's just an SPI port.... How would using the "wrong" rate cause unrecoverable damage? If the port is clocked at 10Mb/s vs 11Mb/s there should be absolutely no visible difference. 

Theoretically if the clock is pushed higher than what the devices support, that could cause bit errors or some such, but that's why I suggested always going lower; never higher.

Reply:
@lsgunth  I agree with Kelvin's observation. 

Normally for SPI clock rate settings, `float` type is more intuitive and a high rate is sufficient. 

However, with the One-Time-Programming (OTP) settings, we want to ensure no ambiguity, and that users understand exactly what they are doing, so `enum` is a perfect way to ensure that. 

How about we add a new function `switchtec_get_spi_clk_str` that converts `enum` to a printable string, as in patch 
684a459 ?

Reply:
I really do not like the enum approach. It was very broken, at doesn't let the user actually specify the SPI rate (which is likely going to produce even worse errors).

If you want to ensure the float during programming is exactly what they specified then do that. Compare the float to the int that was looked up in the table and error out if it's not a valid rate.... However then you make it really hard to use as the user doesn't know all the valid rates.  Ensure you print an error with the nearest valid rate or some such.

Though, I really don't think it's as big an issue as you make it out to be.

Reply:
Yes, the user interface was broken. The ability to add backward compatibility was broken. And it's a cumbersome interface.

The float interface was so much better for all these reasons. Keep it. If you want to enforce the user to pick a supported baudrate and error out when they don't pick a valid value, then do that (but make sure you communicate the valid values). But don't go back to the enum which was the wrong approach from the start.

The whole idea that the exact SPI rate will be a problem is so overblown. It's not like a UART where both sides have to have the same rate. SPI is synchronous so the master can use whatever baudrate it feels like as long as it's not higher than what the other side supports. The exact rate is not important.

Reply:
@lsgunth I agree float would be more intuitive. But I still don't get why `enum` is broken and blocks future backward compatibility. Could you elaborate? I believe an agreement on this would save tons of time in the future reviews.

Reply:
Well, how does the user know what values they can enter?

The original implementation was broken for two reasons: there was no way to change the enum values because they were hard coded outside the library. Second there was no way for the user to know how to set the values because the interface took only a number with no way to know what each number meant. 

Using an enum means there will need to be a bunch of work to map the enumeration values to actual floating point values. Both for printing and for taking input from the user. We'd need an argconfig choices map that differs depending on the depending on the device but can't print help unless it connects to the device. Doing this in such a way that allows for future expansion (by only changing the library) pretty much impossible. Also imagine someone using the library directly without the CLI, they have to repeat this huge pain that then needs to be reworked if the library or enum values need to change or be expanded.

Better to just use a float with some clear methods to determine if a given number is valid. And perhaps a way to list valid floating point numbers. Most users should not care about the exact SPI rate, they just want it to work -- and they especially don't want what they are doing to break if they upgrade their switch.

Reply:
@lsgunth Thanks for the suggestions. 

I have reverted the code back to `float`, but added checking in `config_set` function to ensure (close to) exact match of the input clock rate, or return error code if not. 

Also added a function to allow user to get available rates, so they can chose the preferred rate from there. 

If nothing else, I will go ahead and merge this.